### ISSUE TITLE
Base url as an argument?

### ISSUE THREAD

**Issue by joneshf** [purescript-hyper/purescript-trout-client#4]
This library works really well when you want to hit the server you're serving from. If you need to hit a server at a different location, it seems impossible. How would you feel about taking a base url as an argument?

Something similar to what `runClientM` does in servant-client: https://www.stackage.org/haddock/lts-9.14/servant-client-0.11/Servant-Client.html#v:runClientM

**Comment by joneshf** [comment]
For more context, I thought about suggesting instead to encode the whole URI as the type, but this breaks down when you need to use the same compiled code in different environments. e.g. If you have a test vs a production server, you can't really have one type for both of those. Or can you?

**Comment by owickstrom** [comment]
I think a configurable base URL is the way to go! I probably haven't added support for it as this whole library was quite experimental. I haven't used it for anything serious. If you'd like to send a PR we can get that in. 👍

**Comment by joneshf** [comment]
Actually, maybe the move here is to not specify and just use the facilities that already exist.

```PureScript
import Data.URI (Authority, URIScheme)

newtype BaseURI
  = BaseURI
    { scheme :: URIScheme
    , authority :: Authority
    }

instance hasClientsBaseURI :: ( HasClients sub subMk
                              )
                              => HasClients (BaseURI :> sub) (BaseURI -> subMk) where
  getClients _ req (BaseURI {authority, scheme}) =
    getClients (Proxy :: Proxy sub) (prependSegment schemeAuthority req)
      where
        schemeAuthority = printScheme scheme <> printAuthority authority

prependSegment :: String -> RequestBuilder -> RequestBuilder
prependSegment segment req =
  req { path = singleton segment <> req.path }
```

Then you ought to be able to say

```PureScript
type ForeignAPI
  = BaseURI
  :> "api"
  :/ "v1"
  :/ "users"
  :/ Capture "id" UserId
  :> Resource (Get User JSON)

user :: forall e. BaseURI -> UserId -> { "GET" :: Aff (ajax :: AJAX | e) User }
user = asClients (Proxy :: Proxy ForeignAPI)

type LocalAPI
  = "settings"
  :/ "users"
  :/ Capture "id" UserId
  :> Resource (Get Settings JSON)

settings :: forall e. UserId -> { "GET" :: Aff (ajax :: AJAX | e) Settings }
settings = asClients (Proxy :: Proxy LocalAPI)
```

What do you think?

**Comment by owickstrom** [comment]
Hmm, interesting take. I think I like it, as it makes things more flexible; you could have two different domain names defined in a composite type, like `cdn.example.com` and `app.example.com`. I guess the negative thing as the user of the `asClients` function need to be concerned with base URIs, but you could partially apply that away. Maybe we should just write a small section in the README for this use case?

**Comment by joneshf** [comment]
Do you mean not add `BaseURI` to the module and provide a cookbook/example?

If so, there's a problem. I'm not sure you can define your own base uri elsewhere. it's an orphan to define `HasClients (MyBaseURI :> sub) (MyBaseURI -> subMk)` outside of `Type.Trout.Client`. We'd need a general `HasClients (sup :> sub) (supMk -> subMk)` or something, then we could define each or the `sup` individually. But that breaks down when you have `Lit a` as `sup`.

**Comment by owickstrom** [comment]
Oh sorry, yeah include `BaseURI` with the instance as you did, and show how to use it in the README. Does that sound OK?

**Comment by joneshf** [comment]
Yep!

**Comment by nsaunders** [comment]
Resolved in #16
